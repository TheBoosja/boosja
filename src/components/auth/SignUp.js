import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { registerUser, authError } from 'actions/auth';
import { Form, Error, Button, Title } from 'components/styled';
import renderField from './renderField';

export class SignUp extends Component {
	constructor(props) {
		super(props);

		this.onFormSubmit = this.onFormSubmit.bind(this);
	}

	componentWillUnmount() {
		this.props.authError('');
	}

	renderAlert() {
		const { error } = this.props;
		return error && <Error>{error}</Error>;
	}

	onFormSubmit({ email, password }) {
		this.props.registerUser({ email, password });
	}

	render() {
		const { handleSubmit } = this.props;

		return (
			<Form onSubmit={handleSubmit(this.onFormSubmit)}>
				<Title>Register a new account</Title>
				<Field
					label="Email"
					name="email"
					type="email"
					autoFocus
					component={renderField}
				/>
				<Field
					label="Password"
					name="password"
					type="password"
					component={renderField}
				/>
				<Field
					label="Confirm Password"
					name="passwordConfirm"
					type="password"
					component={renderField}
				/>
				<Button large type="submit">
					Register
				</Button>
				{this.renderAlert()}
			</Form>
		);
	}
}

export function validate({ email, password, passwordConfirm }) {
	const errors = {};

	// Email
	if (!email) {
		errors.email = 'You must enter an email!';
	} else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(email)) {
		errors.email = 'Enter a valid email, please!';
	}

	// Passwords match?
	if (password !== passwordConfirm) {
		errors.passwordConfirm = "Passwords don't match";
	}

	// Password
	if (!password) {
		errors.password = 'You must enter a password';
	} else if (password.length < 8) {
		errors.password = 'The password must contain 8+ characters.';
	}

	// Password Confirmation
	if (!passwordConfirm) {
		errors.passwordConfirm = 'Please re-enter your password!';
	}

	return errors;
}

function mapStateToProps({ auth: { error } }) {
	return { error };
}

export default reduxForm({
	form: 'signup',
	validate
})(connect(mapStateToProps, { registerUser, authError })(SignUp));
