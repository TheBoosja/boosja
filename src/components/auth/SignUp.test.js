import React from 'react';
import { shallow } from 'enzyme';
import { SignUp, validate } from './SignUp';

describe('Sign Up', () => {
	const mockHandleSubmit = jest.fn();
	const mockRegisterUser = jest.fn();
	const mockAuthError = jest.fn();

	const props = { 
		handleSubmit: mockHandleSubmit,
		registerUser: mockRegisterUser, 
		authError: mockAuthError
	};
	const register = shallow(<SignUp {...props} />);

	it('renders properly', () => {
		expect(register).toMatchSnapshot();
	});

	it('contains a Form element', () => {
		expect(register.find('Form').exists()).toBe(true);
	});

	it('contains a Field of type email', () => {
		expect(register.find({ type: 'email' }).exists()).toBe(true);
	});

	it('contains a Field of type password', () => {
		expect(register.find({ type: 'password' }).exists()).toBe(true);
	});

	it('contains a Field of type password and name of `passwordConfirm`', () => {
		expect(register.find({ type: 'password', name: 'passwordConfirm' }).exists()).toBe(true);
	});

	it('contains a submit Button', () => {
		expect(register.find('Button').props().type).toEqual('submit');
	});

	describe('when the user submits the Register from', () => {
		it('calls the `registerUser()` with email and password', () => {
			const cred = { email: 'test@mail.com', password: '1234' };
			register.instance().onFormSubmit(cred);

			expect(mockRegisterUser).toHaveBeenCalledWith(cred);
		});
	});

	describe('when component will unmount', () => {
		it('resets any error from server', () => {
			register.instance().componentWillUnmount();
			expect(mockAuthError).toHaveBeenCalledWith('');
		});
	});

	describe('when an error doesn\'t occur', () => {
		it('does not show an error', () => {
			expect(register.find('Error').exists()).toBe(false);
		});
	});

	describe('when an error occurs', () => {
		it('shows the error', () => {
			const error = { error: 'An error occurred. Do something!' };
			register.setProps(error);

			expect(register.find('Error').children().text()).toEqual(error.error);
		});
	});

	describe('validation', () => {
		describe('when given no credentials', () => {
			const cred = { email: '', password: '', passwordConfirm: '' };
			const res = validate(cred);

			it('email fails', () => {
				expect(res.email).toBeDefined();
			});

			it('password fails', () => {
				expect(res.password).toBeDefined();
			});

			it('password confirmation fails', () => {
				expect(res.passwordConfirm).toBeDefined();
			});
		});

		describe('when given wrong credentials', () => {
			const cred = { email: '123.com', password: '1234', passwordConfirm: '4321' };
			let res = validate(cred);

			describe('email fails', () => {
				it('when no `@` is given', () => {
					expect(res.email).toBeDefined();
				});

				it('when no domain name is given', () => {
					cred.email = '123@';
					res = validate(cred);

					expect(res.email).toBeDefined();
				});

				it('when no ending is given', () => {
					cred.email = '123@mail.';
					res = validate(cred);

					expect(res.email).toBeDefined();
				});
			});

			describe('password fails', () => {
				it('when shorter than 8 characters', () => {
					expect(res.password).toBeDefined();
				});
			});

			describe('password confirmation fails', () => {
				it('when not matching password', () => {
					expect(res.passwordConfirm).toBeDefined();
				});
			});
		});

		describe('when given correct credentials', () => {
			const cred = { email: '123@mail.com', password: '12345678', passwordConfirm: '12345678' };
			const res = validate(cred);

			it('email succeeds', () => {
				expect(res.email).toBeUndefined();
			});

			it('password succeeds', () => {
				expect(res.password).toBeUndefined();
			});

			it('password confirmation succeeds', () => {
				expect(res.passwordConfirm).toBeUndefined();
			});
		});
	});
});