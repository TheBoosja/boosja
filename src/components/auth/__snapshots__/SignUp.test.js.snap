// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp
    authError={[Function]}
    handleSubmit={[Function]}
    registerUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Title>
          Register a new account
</Title>,
        <Field
          autoFocus={true}
          component={[Function]}
          label="Email"
          name="email"
          type="email"
/>,
        <Field
          component={[Function]}
          label="Password"
          name="password"
          type="password"
/>,
        <Field
          component={[Function]}
          label="Confirm Password"
          name="passwordConfirm"
          type="password"
/>,
        <Button
          type="submit"
>
          Register
</Button>,
        undefined,
      ],
      "onSubmit": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Register a new account",
        },
        "ref": null,
        "rendered": "Register a new account",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoFocus": true,
          "component": [Function],
          "label": "Email",
          "name": "email",
          "type": "email",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "label": "Password",
          "name": "password",
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "component": [Function],
          "label": "Confirm Password",
          "name": "passwordConfirm",
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Register",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Register",
        "type": [Function],
      },
      undefined,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Title>
            Register a new account
</Title>,
          <Field
            autoFocus={true}
            component={[Function]}
            label="Email"
            name="email"
            type="email"
/>,
          <Field
            component={[Function]}
            label="Password"
            name="password"
            type="password"
/>,
          <Field
            component={[Function]}
            label="Confirm Password"
            name="passwordConfirm"
            type="password"
/>,
          <Button
            type="submit"
>
            Register
</Button>,
          undefined,
        ],
        "onSubmit": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Register a new account",
          },
          "ref": null,
          "rendered": "Register a new account",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoFocus": true,
            "component": [Function],
            "label": "Email",
            "name": "email",
            "type": "email",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "label": "Password",
            "name": "password",
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "component": [Function],
            "label": "Confirm Password",
            "name": "passwordConfirm",
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Register",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Register",
          "type": [Function],
        },
        undefined,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
